### PR TITLE
GP from AIMD + Serialization

### PR DESCRIPTION
**Pull request by stevetorr**
I spoke with @jonpvandermause  about grandfathering in this commit since I started working on it at the same time as when we resolved to do branch-based development and the commits are a little 'entangled' between changes. In the future I'll split up the individual features on a branched basis, but as is, this PR addresses:

Completely:
- Issue 15: https://github.com/mir-group/flare/issues/15
- Issue 14: https://github.com/mir-group/flare/issues/14  (Though the predict methods are still in OTF.py, we can delete them and refactor OTF later, as they are currently only unit tested in a limited way in OTF and on their own).

Partially:
-  Issue 20 https://github.com/mir-group/flare/issues/20
- Issue 16: https://github.com/mir-group/flare/issues/16

In summary, this PR has:

- A first version of the GP to AIMD module (with 95% unit test coverage!)
- Serialization methods for env and struc
- A JSON encoder to assist with the above serialization methods
- Added/ updated docstrings for many different classes
- PEP8 Compliance (@jonpvandermause , our editors disagree about over / under indentation but everything else should be good)

Feedback welcome:

@nw13slx  Can you weigh in on what the issue was with the predicting while in parallel? I know that you mentioned as much during the meeting last week but I forget what the exact cause was; and, if so, if the current way of splitting it in predict.py  would fix your issues.

### CODE REVIEW

File: flare/gp.py [hunk 1]
```diff
@@ -228,6 +264,13 @@ def en_kern_vec(self, x: AtomicEnvironment) -> np.ndarray:
         return k_v
 
     def set_L_alpha(self):
+        """
+        Invert the covariance matrix, setting L (a lower triangular
+        matrix s.t. L L^T = (K + sig_n^2 I) ) and alpha, an inversion of the
+        vector of training labels.
```
Comment (jonpvandermause) [inline flare/gp.py]: Alpha is the inverse of the covariance matrix multiplied by the vector of training labels.

File: flare/gp_from_aimd.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+                            dt=0,
+                            Nsteps=len(self.frames),
+                            structure=self.frames[0],
+                            std_tolerance=(self.rel_std_tolerance,
+                                           self.abs_std_tolerance),
+                            output_name=self.output_name)
+
+        self.start_time = time.time()
+
+    def run(self):
```
Comment (jonpvandermause) [inline flare/gp_from_aimd.py]: Down the line we'll want to add a number of options for training on a fixed AIMD set; might make sense to generalize the code a little now to make way for future implementations. Passing over each frame and adding environments with high uncertainty makes sense and is data efficient, but it's not the only way (or the easiest). We should also make it easy to train on equally spaced frames, user specified frames, etc. @nw13slx has been playing with different approaches --- it would be good to coordinate with her on this.

File: flare/gp_from_aimd.py [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+    def train_gp(self):
+        self.gp.train(monitor=True if self.verbose >= 2 else False)
+
+        output.write_hyps(self.gp.hyp_labels, self.gp.hyps,
+                          self.start_time, self.output_name,
+                          self.gp.like, self.gp.like_grad)
+        self.train_count += 1
+
+    def is_std_in_bound(self, frame):
```
Comment (jonpvandermause) [inline flare/gp_from_aimd.py]: Looks similar to the is_std_in_bound method of otf. To avoid code duplication, we should have it as a function in one place (either otf or a separate file) and import from there.

File: flare/predict.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+Helper functions which obtain forces and energies
+corresponding to atoms in structures
+"""
+from flare.env import AtomicEnvironment
+from flare.gp import GaussianProcess
+from flare.struc import Structure
+
+import numpy as np
+import concurrent.futures
+
```
Comment (jonpvandermause) [inline flare/predict.py]: These methods are similar to the ones in otf.py, so this file duplicates a lot of code. Before modifying otf.py to call these functions, we'll need to check that they work properly on 32 core jobs (parallelization has been finicky in the past)

File: flare/gp.py [hunk 5]
```diff
@@ -72,6 +74,28 @@ def update_db(self, struc: Structure, forces: list,
         # create numpy array of training labels
         self.training_labels_np = self.force_list_to_np(self.training_labels)
 
+        if self.l_mat is None:
+            self.set_L_alpha()
+        else:
+            self.update_L_alpha()
```
[the remark this answers is not in the document]
Comment (stevetorr) [reply flare/gp.py]: Oh, didn't know update wasn't unit tested. It now will use only set_L_alpha().

File: flare/gp.py (hunk 1 above)
Comment (stevetorr) [reply flare/gp.py]: Whoops, thanks. Fixed!

File: flare/gp_from_aimd.py (hunk 2 above)
Comment (stevetorr) [reply flare/gp_from_aimd.py]: These points are all totally valid. 

This comment touches on the tradeoff to be made here between how much power to cede to the user and how much to build in as functionality, which then results in a more complex training manager. I am pretty conflicted about this myself, and I'm not sure where the line should be. I would love to solicit input from you + @nw13slx + everyone else at a developer meeting about this.

It feels simultaneously like a huge cop out and not a cop out to say that the user can do some of this already. 

For example, if the typehint is followed, one can pass in a list of Structures sliced by e.g.
`input_list[::10]` (though that will not check the MAE against every frame in the trajectory, which might be desirable). An earlier version of the TrajectoryTrainer I had allowed 'seed structures' to be passed in, which would then be applied to the GP before the run through the entire trajectory began. I decided against leaving it in, hoping that the ability to add environments one-by-one into the GP added via the add_one_env method would enable the user to do this before putting their GP in.

On the other perspective, in which the above argument is a cop out, the whole point of TrajectoryTrainer is to automate labor and pre-processing. Musing about where 'the line' is, I think the features you request are probably on the side of inclusion. We can add a 'skip through' option and some version of 'seed frames' which can be passed to the TrajectoryTrainer. One can also do the frames pre-processing or GP pre-processing if they like, and skip the above two features anyway.

File: flare/gp_from_aimd.py (hunk 3 above)
Comment (stevetorr) [reply flare/gp_from_aimd.py]: Agreed. I can transfer this to a different file and re-factor OTF to point there (note that this implementation borrows the good idea of @nw13slx to allow a relative and absolute cutoff simultaneously, and I unit-tested every permutation). I think a good home could be within  predict.py, as the functionality it serves is to be a diagnostic to a prediction made on a given atom.

File: flare/predict.py (hunk 4 above)
Comment (stevetorr) [reply flare/predict.py]: Sounds good and I agree. We can consult @nw13slx  and @YuuuuXie ; I know at least Lixin has worked with an implementation similar to this on her branch.
